### PR TITLE
Pr create https wss endpoints on multiple regions

### PR DESCRIPTION
**Pull request by AndyBoWu**

### CODE REVIEW

File: devops/create_https_wss_endpoints_aws/main.py [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+
+def register_explorers():
+    """
+    register explorer nodes into the corresponding target group
+        * register the same target into tg-https and tg-wss
+    depends on:
+        * dict_region_tgarn
+        *
+    """
+    for i in range(len(ARRAY_OF_REGIONS)):
```
Comment (fxfactorial) [inline devops/create_https_wss_endpoints_aws/main.py]: Range on array of regions directly , use enumerate if really need the index

File: devops/create_https_wss_endpoints_aws/main.py [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+                        Targets=[
+                            {
+                                'Id': instance,
+                                'Port': 9800,
+                            },
+                        ]
+                    )
+                    print("--registering an explorer node into TWO target groups (tg-https and tg-wss) in region " + region)
+                    sleep(4)
+                except Exception as e:
```
Comment (fxfactorial) [inline devops/create_https_wss_endpoints_aws/main.py]: Exception is too base a class to just catch blindly on. Either except on specific exceptions like something from boto or remove the try except and let it explode

File: devops/create_https_wss_endpoints_aws/devnet.json [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  "us-east-2-exp-3": ["3.133.130.192", "18.189.182.113"],
+  "us-west-1-exp-0": ["13.52.243.58","18.144.99.201"],
+  "us-west-1-exp-1": ["18.144.81.143","54.67.74.138"],
+  "us-west-1-exp-2": ["18.144.4.73","13.57.23.204"],
+  "us-west-1-exp-3": ["54.183.244.215","13.57.253.75"],
+  "us-west-2-exp-0": [],
+  "us-west-2-exp-1": [],
+  "us-west-2-exp-2": [],
+  "us-west-2-exp-3": [],
+  "explorers_0" : ["54.191.139.12", "54.71.27.215"],
```
Comment (LeoHChen) [inline devops/create_https_wss_endpoints_aws/devnet.json]: why two IPs?

what's explorer_0, explorer_1?

File: devops/create_https_wss_endpoints_aws/devnet.json [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  "explorers_1" : ["54.241.155.180"],
+  "explorers_2" : ["54.91.183.211"],
+  "tg_s0" : ["tg-s0-api-b-https", "tg-s0-api-b-wss"],
```
Comment (LeoHChen) [inline devops/create_https_wss_endpoints_aws/devnet.json]: is it just a name or what? or pre-defined target group?  If it is just a name generated, why add it to the config file?

File: devops/create_https_wss_endpoints_aws/devnet.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  "tg_s1" : ["tg-s1-api-b-https", "tg-s1-api-b-wss"],
+  "tg_s2" : ["tg-s2-api-b-https", "tg-s2-api-b-wss"],
+  "tg_s3" : ["tg-s3-api-b-https", "tg-s3-api-b-wss"],
+  "ws_endpoints": ["ws.s0.b.hmny.io", "ws.s1.b.hmny.io", "ws.s2.b.hmny.io","ws.s3.b.hmny.io"]
```
Comment (LeoHChen) [inline devops/create_https_wss_endpoints_aws/devnet.json]: what's the relation between the end points and the above values like tg, explorer?  this json config file has to be structured, it is flat right now and hard to figure out the relation among those values.

File: devops/create_https_wss_endpoints_aws/main.py [hunk 6]
```diff
@@ -0,0 +1,520 @@
+#!/usr/bin/env python3
+
+'''
+
+PURPOSE:
+
+create https/wss endpoints for a network
```
Comment (LeoHChen) [inline devops/create_https_wss_endpoints_aws/main.py]: Is this scrip always used to create end points from scratch?  Can it be used to update existing end points, such like we launched testnet multiple times? If yes, how?  Please add it in the usage.

File: devops/create_https_wss_endpoints_aws/main.py [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+
+# store name of target group, key: tg_https, tg_wss, value: array of target group
+dict_tg_https_wss = defaultdict(list)
+
+
+# TO-DO: move to a helper class later
+def parse_network_config(param):
+    """ load the network configuration file, retrieve the value by its key """
+    with open(network_config, 'r') as f:
+        network_config_dict = json.load(f)
```
Comment (LeoHChen) [inline devops/create_https_wss_endpoints_aws/main.py]: it is easy to convert json to dict in python, so you'd consider a bit more structure data.

File: devops/create_https_wss_endpoints_aws/devnet.json [hunk 8]
```diff
@@ -0,0 +1,43 @@
+{
+  "network": "devnet",
+  "domain_name": "b.hmny.io",
+  "num_of_shards": 4,
+  "num_of_regions": 4,
+  "regions": ["us-east-2", "us-west-1"],
+  "region_vpc": ["vpc-2c420a44","vpc-bb770fdc"],
+  "regions_prod": ["us-east-1","us-east-2","us-west-1"],
```
[the remark this answers is not in the document]
Comment (AndyBoWu) [reply devops/create_https_wss_endpoints_aws/devnet.json]: `regions` is the region we would like to create endpoints on. 

`regions_prod` is a reference, cuz it matches the vpc id in the region. Usually, we won't touch this key-value pair.

File: devops/create_https_wss_endpoints_aws/devnet.json (hunk 3 above)
Comment (AndyBoWu) [reply devops/create_https_wss_endpoints_aws/devnet.json]: `explorer_0` has been deprecated. It was created for TG updating. 
There are usually 2 explorers for each endpoint for HA reason.

File: devops/create_https_wss_endpoints_aws/devnet.json (hunk 4 above)
Comment (AndyBoWu) [reply devops/create_https_wss_endpoints_aws/devnet.json]: deprecated design

File: devops/create_https_wss_endpoints_aws/main.py (hunk 6 above)
Comment (AndyBoWu) [reply devops/create_https_wss_endpoints_aws/main.py]: yes, it is described in the __main__ func.

File: devops/create_https_wss_endpoints_aws/main.py (hunk 1 above)
Comment (AndyBoWu) [reply devops/create_https_wss_endpoints_aws/main.py]: could do that too

**Comment by AndyBoWu** [comment]
changes on feb 1, 2020

1. removed NUM_OF_REGIONS from json file and main.py, redundant
2. renamed `regions_prod` to `regions_ref`; renamed `region_vpc_prod` to `region_vpc_ref`
3. removed `regions_code` from json, deprecated
4. code refactoring for the ssl cert requesting

**Pull request merged**